File: Example/FWFramework/Classes/Module/Test/Plugin/TestToastController.m
//
//  TestToastController.m
//  FWFramework_Example
//
//  Created by wuyong on 2022/8/25.
//  Copyright © 2022 CocoaPods. All rights reserved.
//

#import "TestToastController.h"
#import "AppSwift.h"
@import FWFramework;

@interface TestToastController () <FWTableViewController>

@end

@implementation TestToastController

- (UITableViewStyle)setupTableStyle
{
    return UITableViewStyleGrouped;
}

- (void)setupSubviews
{
    [self.tableData addObjectsFromArray:@[
        @[@"无文本", @"onIndicator"],
        @[@"有文本(可取消)", @"onIndicator2"],
        @[@"文本太长", @"onIndicator3"],
        @[@"加载动画", @"onLoading"],
        @[@"进度动画(可取消)", @"onProgress"],
        @[@"加载动画(window)", @"onLoadingWindow"],
        @[@"加载进度动画(window)", @"onProgressWindow"],
        @[@"单行吐司(可点击)", @"onToast"],
        @[@"多行吐司(默认不可点击)", @"onToast2"],
    ]];
}

#pragma mark - TableView

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section
{
    return self.tableData.count;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath
{
    UITableViewCell *cell = [UITableViewCell fw_cellWithTableView:tableView];
    NSArray *rowData = [self.tableData objectAtIndex:indexPath.row];
    cell.textLabel.text = [rowData objectAtIndex:0];
    return cell;
}

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath
{
    [tableView deselectRowAtIndexPath:indexPath animated:YES];
    NSArray *rowData = [self.tableData objectAtIndex:indexPath.row];
    SEL selector = NSSelectorFromString([rowData objectAtIndex:1]);
    if ([self respondsToSelector:selector]) {
        FWIgnoredBegin();
        [self performSelector:selector];
        FWIgnoredEnd();
    }
}

#pragma mark - Action

- (void)onIndicator
{
    [self fw_showLoadingWithText:@""];
    dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(2.0 * NSEC_PER_SEC)), dispatch_get_main_queue(), ^{
        [self fw_hideLoading];
    });
}

- (void)onIndicator2
{
    static BOOL isCancelled = NO;
    FWWeakifySelf();
    isCancelled = NO;
    [self fw_showLoadingWithText:nil cancelBlock:^{
        isCancelled = YES;
        FWStrongifySelf();
        [self fw_showMessageWithText:@"已取消"];
    }];
    dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(2.0 * NSEC_PER_SEC)), dispatch_get_main_queue(), ^{
        if (isCancelled) return;
        FWStrongifySelf();
        [self fw_hideLoading];
    });
}

- (void)onIndicator3
{
    [self fw_showLoadingWithText:@"我是很长很长很长很长很长很长很长很长很长很长的加载文案"];
    dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(2.0 * NSEC_PER_SEC)), dispatch_get_main_queue(), ^{
        [self fw_hideLoading];
    });
}

- (void)onLoading
{
    [self fw_showLoadingWithText:@"加载中\n请耐心等待"];
    dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(2.0 * NSEC_PER_SEC)), dispatch_get_main_queue(), ^{
        [self fw_hideLoading];
    });
}

- (void)onProgress
{
    static BOOL isCancelled = NO;
    FWWeakifySelf();
    isCancelled = NO;
    [TestController mockProgress:^(double progress, BOOL finished) {
        if (isCancelled) return;
        FWStrongifySelf();
        if (!finished) {
            [self fw_showProgressWithText:[NSString stringWithFormat:@"上传中(%.0f%%)", progress * 100] progress:progress cancelBlock:^{
                isCancelled = YES;
                FWStrongifySelf();
                [self fw_showMessageWithText:@"已取消"];
            }];
        } else {
            [self fw_hideProgress];
        }
    }];
}

- (void)onLoadingWindow
{
    self.view.window.fw_toastInsets = UIEdgeInsetsMake(FWTopBarHeight, 0, 0, 0);
    [self.view.window fw_showLoadingWithText:@"加载中"];
    dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(2.0 * NSEC_PER_SEC)), dispatch_get_main_queue(), ^{
        self.view.window.fw_toastInsets = UIEdgeInsetsZero;
        [self.view.window fw_hideLoading];
    });
}

- (void)onProgressWindow
{
    FWWeakifySelf();
    [TestController mockProgress:^(double progress, BOOL finished) {
        FWStrongifySelf();
        if (!finished) {
            [self.view.window fw_showLoadingWithText:[NSString stringWithFormat:@"上传中(%.0f%%)", progress * 100]];
        } else {
            [self.view.window fw_hideLoading];
        }
    }];
}

- (void)onToast
{
    self.view.tag = 100;
    static int count = 0;
    [self fw_showMessageWithText:[NSString stringWithFormat:@"吐司消息%@", @(++count)]];
}

- (void)onToast2
{
    NSString *text = @"我是很长很长很长很长很长很长很长很长很长很长很长的吐司消息";
    FWWeakifySelf();
    [self fw_showMessageWithText:text style:FWToastStyleDefault completion:^{
        FWStrongifySelf();
        [self onToast];
    }];
}

@end

File: Example/FWFramework/Classes/Module/Test/Plugin/TestToastController.swift
//
//  TestToastController.swift
//  FWFramework_Example
//
//  Created by wuyong on 2022/9/15.
//  Copyright © 2022 CocoaPods. All rights reserved.
//

import FWFramework

class TestToastController: UIViewController, TableViewControllerProtocol {
    
    typealias TableElement = [String]
    
    static var isCancelled = false
    static var count: Int = 0
    
    func setupTableStyle() -> UITableView.Style {
        .grouped
    }
    
    func setupSubviews() {
        tableData.append(contentsOf: [
            ["无文本", "onIndicator"],
            ["有文本(可取消)", "onIndicator2"],
            ["文本太长", "onIndicator3"],
            ["加载动画", "onLoading"],
            ["加载动画(连续)", "onContinuous"],
            ["进度动画(可取消)", "onProgress"],
            ["加载动画(window)", "onLoadingWindow"],
            ["加载进度动画(window)", "onProgressWindow"],
            ["单行吐司(可点击)", "onToast"],
            ["多行吐司(默认不可点击)", "onToast2"],
        ])
    }
    
    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        tableData.count
    }
    
    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cell = UITableViewCell.app.cell(tableView: tableView)
        let rowData = tableData[indexPath.row]
        cell.textLabel?.text = rowData[0]
        return cell
    }
    
    func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        tableView.deselectRow(at: indexPath, animated: true)
        let rowData = tableData[indexPath.row]
        _ = self.perform(NSSelectorFromString(rowData[1]))
    }
    
}

@objc extension TestToastController {
    
    func onIndicator() {
        app.showLoading(text: "")
        DispatchQueue.main.asyncAfter(deadline: .now() + 2) { [weak self] in
            self?.app.hideLoading()
        }
    }
    
    func onIndicator2() {
        Self.isCancelled = false
        app.showLoading(text: nil) { [weak self] in
            Self.isCancelled = true
            self?.app.showMessage(text: "已取消")
        }
        DispatchQueue.main.asyncAfter(deadline: .now() + 2) { [weak self] in
            if Self.isCancelled { return }
            self?.app.hideLoading()
        }
    }
    
    func onIndicator3() {
        app.showLoading(text: "我是很长很长很长很长很长很长很长很长很长很长的加载文案")
        DispatchQueue.main.asyncAfter(deadline: .now() + 2) { [weak self] in
            self?.app.hideLoading()
        }
    }
    
    func onLoading() {
        app.showLoading(text: "加载中\n请耐心等待")
        DispatchQueue.main.asyncAfter(deadline: .now() + 2) { [weak self] in
            self?.app.hideLoading()
        }
    }
    
    func onContinuous() {
        app.showLoading(text: "加载1中")
        DispatchQueue.main.asyncAfter(deadline: .now() + 2) { [weak self] in
            self?.app.hideLoading(delayed: true)
            
            self?.app.showLoading(text: "加载2中")
            DispatchQueue.main.asyncAfter(deadline: .now() + 2) { [weak self] in
                self?.app.hideLoading()
            }
        }
    }
    
    func onProgress() {
        Self.isCancelled = false
        TestController.mockProgress { [weak self] progress, finished in
            if Self.isCancelled { return }
            if !finished {
                self?.app.showProgress(progress, text: String(format: "上传中(%.0f%%)", progress * 100), cancelBlock: {
                    Self.isCancelled = true
                    self?.app.showMessage(text: "已取消")
                })
            } else {
                self?.app.hideProgress()
            }
        }
    }
    
    func onLoadingWindow() {
        view.window?.app.toastInsets = UIEdgeInsets(top: APP.topBarHeight, left: 0, bottom: 0, right: 0)
        view.window?.app.showLoading(text: "加载中")
        DispatchQueue.main.asyncAfter(deadline: .now() + 2) { [weak self] in
            self?.view.window?.app.toastInsets = .zero
            self?.view.window?.app.hideLoading()
        }
    }
    
    func onProgressWindow() {
        TestController.mockProgress { [weak self] progress, finished in
            if !finished {
                self?.view.window?.app.showLoading(text: String(format: "上传中(%.0f%%)", progress * 100))
            } else {
                self?.view.window?.app.hideLoading()
            }
        }
    }
    
    func onToast() {
        view.tag = 100
        Self.count += 1
        app.showMessage(text: "吐司消息\(Self.count)")
    }
    
    func onToast2() {
        app.showMessage(text: "我是很长很长很长很长很长很长很长很长很长很长很长的吐司消息", style: .default) { [weak self] in
            self?.onToast()
        }
    }
    
}
